File: algorithms/132.Palindrome_Partitioning_II/Palindrome_Partitioning_II.java
import java.util.*;
import java.util.regex.Matcher;

public class Solution {
    public int minCut(String s) {
        int n = s.length();
        boolean[][] isPalindr = new boolean[n + 1][n + 1];
        int[] dp = new int[n + 1];

        // init
        for (int i = 0; i <= n; ++i) {
            dp[i] = i - 1;
        }

        // dp
        for (int i = 2; i <= n; ++i) {
            for (int j = i - 1; j >= 0; --j) {
                if (s.charAt(i - 1) == s.charAt(j) &&
                        (i - 1- j < 2 || isPalindr[j + 1][i - 1])) {
                    isPalindr[j][i] = true;
                    dp[i] = Math.min(dp[i], dp[j] + 1);
                }
            }
        }

        return dp[n];
    }

    // debug
    public static void main(String[] args) {
        Scanner cin = new Scanner(System.in);
        Solution s = new Solution();
        int[] input = {1, 2, 3, 4};
        System.out.println(s.minCut("ababcaca"));
    }
}
public class Solution {
    private boolean isPalindrome(String s, int start, int end) {
        for (int i = start, j = end; i < j; i++, j--) {
            if (s.charAt(i) != s.charAt(j)) {
                return false;
            }
        }
        return true;
    }

    private boolean[][] getIsPalindrome(String s) {
        boolean[][] isPalindrome = new boolean[s.length()][s.length()];

        for (int i = 0; i < s.length(); i++) {
            isPalindrome[i][i] = true;
        }
        for (int i = 0; i < s.length() - 1; i++) {
            isPalindrome[i][i + 1] = (s.charAt(i) == s.charAt(i + 1));
        }

        for (int length = 2; length < s.length(); length++) {
            for (int start = 0; start + length < s.length(); start++) {
                isPalindrome[start][start + length]
                    = isPalindrome[start + 1][start + length - 1] && s.charAt(start) == s.charAt(start + length);
            }
        }

        return isPalindrome;
    }

    public int minCut(String s) {
        if (s == null || s.length() == 0) {
            return 0;
        }

        // preparation
        boolean[][] isPalindrome = getIsPalindrome(s);
        
        // initialize
        int[] f = new int[s.length() + 1];
        for (int i = 0; i <= s.length(); i++) {
            f[i] = i - 1;
        }
        
        // main
        for (int i = 1; i <= s.length(); i++) {
            for (int j = 0; j < i; j++) {
                if (isPalindrome[j][i - 1]) {
                    f[i] = Math.min(f[i], f[j] + 1);
                }
            }
        }

        return f[s.length()];
    }
}

File: algorithms/132.Palindrome_Partitioning_II/Palindrome_Partitioning_II.cpp
<<<<<<< HEAD
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int minCut(string s) {
        int n = s.length();
        vector<int> dp(n + 1);

        // init
        for (int i = 0; i <= n; ++i)
            dp[i] = i - 1;

        // dp
        for (int i = 0; i < n; ++i) {
            // odd
            for (int j = 0; i - j >= 0 && i + j < n
                    && s[i - j] == s[i + j]; ++j)
                dp[i + j + 1] = min(dp[i + j + 1], dp[i - j] + 1);
            // even
            for (int j = 1; i - j + 1 >= 0 && i + j < n
                    && s[i - j + 1] == s[i + j]; ++j)
                dp[i + j + 1] = min(dp[i + j + 1], dp[i - j + 1] + 1);
        }

        return dp[n];
    }
};

int main() {

    return 0;
}

=======
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int minCut(string s) {
        int n = s.length();
        vector<int> dp(n + 1);

        // init
        for (int i = 0; i <= n; ++i)
            dp[i] = i - 1;

        // dp
        for (int i = 0; i < n; ++i) {
            // odd
            for (int j = 0; i - j >= 0 && i + j < n
                    && s[i - j] == s[i + j]; ++j)
                dp[i + j + 1] = min(dp[i + j + 1], dp[i - j] + 1);
            // even
            for (int j = 1; i - j + 1 >= 0 && i + j < n
                    && s[i - j + 1] == s[i + j]; ++j)
                dp[i + j + 1] = min(dp[i + j + 1], dp[i - j + 1] + 1);
        }

        return dp[n];
    }
};

int main() {

    return 0;
}

>>>>>>> 6200c8704614e918c8bfa5357c648dd1b4f7eb74
